src/commands/stash.js

```diff
@@ -0,0 +1,201 @@
+// @ts-check
+import '../typedefs.js'
+
+import { checkout } from '../api/checkout.js'
+import { readCommit } from '../api/readCommit.js'
+import { NotFoundError } from '../errors/NotFoundError.js'
+import { GitRefManager } from '../managers/GitRefManager.js'
+import { GitStashManager } from '../managers/GitStashManager.js'
+import {
+  writeTreeChanges,
+  applyTreeChanges,
+  acquireLock,
+} from '../utils/walkerToTreeEntryMap.js'
+
+import { STAGE } from './STAGE.js'
+import { TREE } from './TREE.js'
+import { _currentBranch } from './currentBranch.js'
+import { _readCommit } from './readCommit.js'
+
+export async function _stashPush({ fs, dir, gitdir, message = '' }) {
+  const stashMgr = new GitStashManager({ fs, dir, gitdir })
+
+  await stashMgr.getAuthor() // ensure there is an author
+  const branch = await _currentBranch({
+    fs,
+    gitdir,
+    fullname: false,
+  })
+
+  // prepare the stash commit: first parent is the current branch HEAD
+  const headCommit = await GitRefManager.resolve({
+    fs,
+    gitdir,
+    ref: 'HEAD',
+  })
+
+  const headCommitObj = await readCommit({ fs, dir, gitdir, oid: headCommit })
+  const headMsg = headCommitObj.commit.message
+
+  const stashCommitParents = [headCommit]
+  let stashCommitTree = null
+  let workDirCompareBase = TREE({ ref: 'HEAD' })
+
+  const indexTree = await writeTreeChanges(fs, dir, gitdir, [
+    TREE({ ref: 'HEAD' }),
+    'stage',
+  ])
+  if (indexTree) {
+    // this indexTree will be the tree of the stash commit
+    // create a commit from the index tree, which has one parent, the current branch HEAD
+    const stashCommitOne = await stashMgr.writeStashCommit({
+      message: `stash-Index: WIP on ${branch} - ${new Date().toISOString()}`,
+      tree: indexTree, // stashCommitTree
+      parent: stashCommitParents,
+    })
+    stashCommitParents.push(stashCommitOne)
+    stashCommitTree = indexTree
+    workDirCompareBase = STAGE()
+  }
+
+  const workingTree = await writeTreeChanges(fs, dir, gitdir, [
+    workDirCompareBase,
+    'workdir',
+  ])
+  if (workingTree) {
+    // create a commit from the working directory tree, which has one parent, either the one we just had, or the headCommit
+    const workingHeadCommit = await stashMgr.writeStashCommit({
+      message: `stash-WorkDir: WIP on ${branch} - ${new Date().toISOString()}`,
+      tree: workingTree,
+      parent: [stashCommitParents[stashCommitParents.length - 1]],
+    })
+
+    stashCommitParents.push(workingHeadCommit)
+    stashCommitTree = workingTree
+  }
+
+  if (!stashCommitTree || (!indexTree && !workingTree)) {
+    throw new NotFoundError('changes, nothing to stash')
+  }
+
+  // create another commit from the tree, which has three parents: HEAD and the commit we just made:
+  const stashCommit = await stashMgr.writeStashCommit({
+    message:
+      message.trim() || `stash: WIP on ${branch} - ${new Date().toISOString()}`,
```

Comment:
Also, the default message for git stash looks like this in canonical git:

```
stash@{13}: WIP on devel: 0e5cff3 fix indent of doc strings
```

It contain hash and last commit message (first line).

Reply:
the prefix is auto generated, I'll update the default message to align with the default